Code/Yospace/Source/YospaceManager.swift

```diff
@@ -0,0 +1,151 @@
+        self.source = nil
+        self.id3MetadataHandler = nil
+        self.playerEventsHandler = nil
+        self.yospaceNotificationsHandler = nil
+    }
+
+    func destroy() {
+        self.reset()
+        self.adIntegrationHandler = nil
+        self.adIntegrationController = nil
+        self.eventDispatcher.clear()
+    }
+
+    private func onSessionCreate(session: YOSession) {
+        let state: YOSessionState = session.sessionState
+        if state == .initialised || state == .noAnalytics {
+            self.setupManager(session: session)
+        } else if state == .failed {
+            let message: String = YOSession.YOSessionResultCode(rawValue: session.resultCode)?.message() ?? "Yospace: Session could not be initialised"
+            let error = YospaceError.error(msg: message)
+            self.adIntegrationController?.fatalError(error: error, code: .AD_ERROR)
+        }
+    }
+
+    private func setupManager(session: YOSession) {
+        if let playbackUrlStr: String = session.playbackUrl,
+           let playbackUrl: URL = .init(string: playbackUrlStr) {
+            self.yospaceSession = session
+            (self.adIntegrationHandler as? YospaceHandler)?.session = session
+            self.id3MetadataHandler = YospaceID3MetadataHandler(player: self.player, session: session)
+            self.playerEventsHandler = THEOplayerEventsHandler(player: self.player, session: session)
+            if let controller: THEOplayerSDK.ServerSideAdIntegrationController = self.adIntegrationController {
+                self.yospaceNotificationsHandler = YospaceNotificationsHandler(session: session, adIntegrationController: controller, player: self.player)
+            }
+            if let source: YospaceManagerSource = self.source {
+                let typedSource: THEOplayerSDK.TypedSource = source.1
+                typedSource.src = playbackUrl
+                let sourceDescription: THEOplayerSDK.SourceDescription = source.0
+                self.isSettingSource = true
+                self.player.source = sourceDescription
+                self.isSettingSource = false
+            }
+            self.yospaceSession?.setPlaybackPolicyHandler(DefaultPlaybackPolicy(playbackMode: session.playbackMode))
+            let event: SessionAvailableEvent = .init(date: Date())
+            self.eventDispatcher.dispatchEvent(event: event)
+        }
+    }
+
+    deinit {
+        self.destroy()
+    }
+}
+
+class YospaceHandler: THEOplayerSDK.ServerSideAdIntegrationHandler {
+    weak var player: THEOplayerSDK.THEOplayer?
+    weak var manager: YospaceManager?
+    var session: YOSession?
+
+    init(player: THEOplayerSDK.THEOplayer, manager: YospaceManager) {
+        self.player = player
+        self.manager = manager
+    }
+
+    func setSource(source: SourceDescription) -> Bool {
+        self.manager?.createYospaceSource(sourceDescription: source, sessionProperties: nil)
+        return true
+    }
```

Comment:
Shouldn't there be an `if self.manager?.isSettingSource == false` check here? Because we are also changing the source in `setupManager()`.

Reply:
Not necessary. The change made to the source in `setupManager()` goes into a different flow and does not reach back into `setSource`.

Reply:
Changed already with `didSetSourceFromConnector`